fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PullUpJoinFromUnion.java

```diff
@@ -0,0 +1,585 @@
+            Plan otherSide;
+            if (pair.second == pair.first.left()) {
+                otherSide = pair.first.right();
+            } else {
+                otherSide = pair.first.left();
+            }
+            List<NamedExpression> projects = otherOutputsList.get(i);
+            // In projects, we also need to add the other slot in join condition
+            Map<SlotReference, List<SlotReference>> commonSlotToOtherSlotMap = commonSlotToOtherSlotMaps.get(i);
+            for (SlotReference commonSlot : joinCommonSlots) {
+                List<SlotReference> otherSlots = commonSlotToOtherSlotMap.get(commonSlot);
+                for (SlotReference otherSlot : otherSlots) {
+                    if (i == 0) {
+                        int index = projects.size();
+                        commonSlotToProjectsIndex.computeIfAbsent(commonSlot, k -> new ArrayList<>()).add(index);
+                    }
+                    projects.add(otherSlot);
+                }
+            }
+            LogicalProject<Plan> logicalProject = new LogicalProject<>(projects, otherSide);
+            newChildren.add(logicalProject);
+        }
+
+        //2. construct new union
+        LogicalUnion newUnion = new LogicalUnion(Qualifier.ALL, newChildren);
+        List<List<SlotReference>> childrenOutputs = newChildren.stream()
+                .map(p -> p.getOutput().stream()
+                        .map(SlotReference.class::cast)
+                        .collect(ImmutableList.toImmutableList()))
+                .collect(ImmutableList.toImmutableList());
+        newUnion = (LogicalUnion) newUnion.withChildrenAndTheirOutputs(newChildren, childrenOutputs);
+        newUnion = newUnion.withNewOutputs(newUnion.buildNewOutputs());
+        return newUnion;
+    }
+
+    // This function is used to check whether the join condition meets the optimization condition
+    // Check the join condition, requiring that the join condition of each join is equal and the number is the same.
+    // Generate commonSlotToOtherSlotMaps. In each map of the list, the keySet must be the same,
+    // and the length of the value list of the same key must be the same.
+    // Output parameter: commonSlotToOtherSlotMaps, which records the join condition of each join.
+    // The key is the slot on the common side of the join, and the value is the slot on the other side of the join.
+    // e.g. select t2.a+1,2 from test_like1 t1 join test_like2 t2 on t1.a=t2.a and t1.a=t2.c and t1.b=t2.b union ALL
+    // select t3.a+1,3 from test_like1 t1 join test_like3 t3 on t1.a=t3.a and t1.a=t3.d and t1.b=t3.b
+    // commonSlotToOtherSlotMaps： {{t1.a:t2.a,t2.c; t1.b:t2.b},{t1.a:t3.a,t3.d; t1.b:t3.b}}
+    // commonSlotToOtherSlotMaps is used to check whether the join condition meets the optimization conditions
+    // and to generate the join condition for the new join
+    private boolean checkJoinCondition(List<Pair<LogicalJoin<?, ?>, Plan>> joinsAndCommonSides,
```

Comment:
1. please give example for all the rejected cases
2. it is better to return Optional<commonSlotToOtherSlotMaps>